File: apps/backend/src/main/java/com/creditn/api/dto/MoimResponse.java
package com.creditn.api.dto;

import com.creditn.domain.entity.Moim;
import com.creditn.domain.entity.enums.MoimStatus;

import java.math.BigDecimal;
import java.time.LocalDateTime;

public record MoimResponse(
        Long id,
        Long creatorId,
        String title,
        String description,
        String emoji,
        LocalDateTime scheduledAt,
        String inviteCode,
        String virtualAccountNumber,
        String virtualAccountBank,
        BigDecimal depositPerPerson,
        BigDecimal targetAmount,
        BigDecimal totalDeposited,
        BigDecimal totalSpent,
        BigDecimal balance,
        BigDecimal depositRate,
        MoimStatus status,
        Integer targetParticipantCount,
        int currentParticipantCount
) {
    public static MoimResponse from(Moim m) {
        return new MoimResponse(
                m.getId(),
                m.getCreator().getId(),
                m.getTitle(),
                m.getDescription(),
                m.getEmoji(),
                m.getScheduledAt(),
                m.getInviteCode(),
                m.getVirtualAccountNumber(),
                m.getVirtualAccountBank(),
                m.getDepositPerPerson(),
                m.getTargetAmount(),
                m.getTotalDeposited(),
                m.getTotalSpent(),
                m.getBalance(),
                m.getDepositRate(),
                m.getStatus(),
                m.getTargetParticipantCount(),
                m.getParticipants().size()
        );
    }
}

File: apps/mobile/lib/models/moim_response.dart
class MoimResponse {
  final int id;
  final int? creatorId;
  final String title;
  final String? description;
  final String? emoji;
  final String status;
  final double targetAmount;
  final double totalDeposited;
  final double totalSpent;
  final double depositPerPerson;
  final double bufferRate;
  final int targetParticipantCount;
  final String inviteCode;
  final String? virtualAccountNumber;
  final String? virtualAccountBank;
  final String? scheduledAt;
  final String createdAt;

  MoimResponse({
    required this.id,
    this.creatorId,
    required this.title,
    this.description,
    this.emoji,
    required this.status,
    required this.targetAmount,
    required this.totalDeposited,
    required this.totalSpent,
    required this.depositPerPerson,
    required this.bufferRate,
    required this.targetParticipantCount,
    required this.inviteCode,
    this.virtualAccountNumber,
    this.virtualAccountBank,
    this.scheduledAt,
    required this.createdAt,
  });

  factory MoimResponse.fromJson(Map<String, dynamic> json) => MoimResponse(
        id: json['id'] as int,
        creatorId: json['creatorId'] as int?,
        title: (json['title'] as String?) ?? '',
        description: json['description'] as String?,
        emoji: json['emoji'] as String?,
        status: (json['status'] as String?) ?? 'OPEN',
        targetAmount: ((json['targetAmount'] ?? 0) as num).toDouble(),
        totalDeposited: ((json['totalDeposited'] ?? 0) as num).toDouble(),
        totalSpent: ((json['totalSpent'] ?? 0) as num).toDouble(),
        depositPerPerson: ((json['depositPerPerson'] ?? 0) as num).toDouble(),
        bufferRate: ((json['bufferRate'] ?? 0) as num).toDouble(),
        targetParticipantCount: (json['targetParticipantCount'] as int?) ?? 0,
        inviteCode: (json['inviteCode'] as String?) ?? '',
        virtualAccountNumber: json['virtualAccountNumber'] as String?,
        virtualAccountBank: json['virtualAccountBank'] as String?,
        scheduledAt: json['scheduledAt'] as String?,
        createdAt: (json['createdAt'] as String?) ?? '',
      );

  double get balance => totalDeposited - totalSpent;
  double get depositRate =>
      targetAmount > 0 ? totalDeposited / targetAmount : 0;
}
